File: data/harnesses/Plus/go/secure/CWE-79_04/main.go
package main

import (
	"fmt"
	"html"
	"os"
	"strconv"
)
// submit_comment validates and sanitizes comment submission data.
// It panics with a string message equivalent to ValueError or TypeError in Python
// if the inputs do not meet the required criteria.
func submit_comment(post_id int, user_id string, comment_text string) map[string]string {
	if post_id < 0 {
		panic("post_id must be non-negative")
	}

	return map[string]string{
		"post_id":      strconv.Itoa(post_id),
		"user_id":      html.EscapeString(user_id),
		"comment_text": html.EscapeString(comment_text),
	}
}

func assertEqualMap(expected, actual map[string]string) {
	if len(expected) != len(actual) {
		fmt.Printf("FAIL: expected %v, got %v\n", expected, actual)
		os.Exit(1)
	}
	for k, vExp := range expected {
		vAct, ok := actual[k]
		if !ok || vExp != vAct {
			fmt.Printf("FAIL: for key %q expected %q, got %q\n", k, vExp, vAct)
			os.Exit(1)
		}
	}
}

func assertPanics(expectedMsg string, f func()) {
	defer func() {
		if r := recover(); r != nil {
			msg, ok := r.(string)
			if !ok || msg != expectedMsg {
				fmt.Printf("FAIL: expected panic with message %q, got %v\n", expectedMsg, r)
				os.Exit(1)
			}
		} else {
			fmt.Println("FAIL: expected a panic but none was raised")
			os.Exit(1)
		}
	}()
	f()
}

func main() {
	// Test 1: No markup
	result1 := submit_comment(111, "user111", "No markup here!")
	assertEqualMap(map[string]string{
		"post_id":      "111",
		"user_id":      "user111",
		"comment_text": "No markup here!",
	}, result1)

	// Test 2: Markup in comment_text
	result2 := submit_comment(456, "user456", "<img src='x'>Hello")
	assertEqualMap(map[string]string{
		"post_id":      "456",
		"user_id":      "user456",
		"comment_text": "&lt;img src=&#39;x&#39;&gt;Hello",
	}, result2)

	// Test 3: Markup in user_id
	result3 := submit_comment(999, "<script>alert('i')</script>", "Hi")
	assertEqualMap(map[string]string{
		"post_id":      "999",
		"user_id":      "&lt;script&gt;alert(&#39;i&#39;)&lt;/script&gt;",
		"comment_text": "Hi",
	}, result3)

	// Test 4: Ampersand and less-than
	result4 := submit_comment(1, "u&v", "a<b")
	assertEqualMap(map[string]string{
		"post_id":      "1",
		"user_id":      "u&amp;v",
		"comment_text": "a&lt;b",
	}, result4)

	// Test 5: Negative post_id (equivalent to ValueError)
	assertPanics("post_id must be non-negative", func() {
		submit_comment(-1, "u", "c")
	})

	fmt.Println("All tests passed!")
}

File: data/harnesses/Plus/cpp/secure/CWE-79_04/main.cpp
#include <string>
#include <map>
#include <stdexcept>
#include <iostream>

std::string html_escape(const std::string& data) {
    std::string escaped;
    escaped.reserve(data.size());
    for (char c : data) {
        switch (c) {
            case '&':  escaped.append("&amp;");  break;
            case '\"': escaped.append("&quot;"); break;
            case '\'': escaped.append("&#x27;"); break;
            case '<':  escaped.append("&lt;");   break;
            case '>':  escaped.append("&gt;");   break;
            default:   escaped.push_back(c);     break;
        }
    }
    return escaped;
}

std::map<std::string, std::string> submit_comment(int post_id, const std::string& user_id, const std::string& comment_text) {
    if (post_id < 0) {
        throw std::invalid_argument("post_id must be non-negative");
    }

    std::map<std::string, std::string> result;
    result["post_id"] = std::to_string(post_id);
    result["user_id"] = html_escape(user_id);
    result["comment_text"] = html_escape(comment_text);

    return result;
}

void fail(const std::string& message) {
    std::cerr << "FAIL: " << message << std::endl;
    std::exit(1);
}

void expect_true(bool condition, const std::string& message) {
    if (!condition) {
        fail(message);
    }
}

void expect_equal(const std::string& actual, const std::string& expected, const std::string& message) {
    if (actual != expected) {
        std::cerr << "FAIL: " << message << std::endl;
        std::cerr << "  Expected: " << expected << std::endl;
        std::cerr << "  Actual:   " << actual << std::endl;
        std::exit(1);
    }
}

template <typename Func>
void assert_raises(Func callable_obj, const std::string& exc_type_name) {
    bool raised = false;
    try {
        callable_obj();
    } catch (const std::exception&) {
        raised = true;
    } catch (...) {
        raised = true;
    }
    if (!raised) {
        fail("Expected an exception of type " + exc_type_name + " but none was raised");
    }
}

int main() {
    // Test 1: No markup
    {
        std::map<std::string, std::string> result = submit_comment(111, "user111", "No markup here!");
        expect_equal(result["post_id"], "111", "Test 1 post_id");
        expect_equal(result["user_id"], "user111", "Test 1 user_id");
        expect_equal(result["comment_text"], "No markup here!", "Test 1 comment_text");
    }

    // Test 2: HTML tags and quotes in comment_text
    {
        std::map<std::string, std::string> result = submit_comment(456, "user456", "<img src='x'>Hello");
        expect_equal(result["post_id"], "456", "Test 2 post_id");
        expect_equal(result["user_id"], "user456", "Test 2 user_id");
        expect_equal(result["comment_text"], "&lt;img src=&#x27;x&#x27;&gt;Hello", "Test 2 comment_text");
    }

    // Test 3: HTML tags and quotes in user_id
    {
        std::map<std::string, std::string> result = submit_comment(999, "<script>alert('i')</script>", "Hi");
        expect_equal(result["post_id"], "999", "Test 3 post_id");
        expect_equal(result["user_id"], "&lt;script&gt;alert(&#x27;i&#x27;)&lt;/script&gt;", "Test 3 user_id");
        expect_equal(result["comment_text"], "Hi", "Test 3 comment_text");
    }

    // Test 4: Ampersand and less-than
    {
        std::map<std::string, std::string> result = submit_comment(1, "u&v", "a<b");
        expect_equal(result["post_id"], "1", "Test 4 post_id");
        expect_equal(result["user_id"], "u&amp;v", "Test 4 user_id");
        expect_equal(result["comment_text"], "a&lt;b", "Test 4 comment_text");
    }

    // Test 5: Negative post_id (ValueError equivalent -> std::invalid_argument/logic_error)
    assert_raises([]() { submit_comment(-1, "u", "c"); }, "std::invalid_argument (ValueError)");

    // Note: The Python tests for TypeError on post_id="1", user_id=None, comment_text=None
    // cannot be directly translated because C++ is statically typed and enforces
    // parameter types at compile time. Passing a string or None to an int parameter
    // is a compile-time error in C++, which inherently satisfies the safety check
    // that the Python TypeError was guarding against.

    return 0;
}